eclipse/openvsx
# Etag and caching

Fixes #391
Contributes to #379

The server uses a fixed caching strategy. `/api/{namespace}/{extension}` responses are cached for 10 minutes and `/api/{namespace}/{extension}/{version}` responses are cached for 7 days. When changing the version using the dropdown, the browser location changes to `/extension/{namespace}/{extension}/{version}`. If the version page has been visited before, the browser loads JSON from cache, else it sends a request to the API. Data inconsistencies emerge when some versions are loaded from cache while others are newly requested and the extension has been updated in the meantime (download count increase, version was added or removed, user details changed).

Instead of caching responses for a fixed duration, use Etag response header to always return the current extension state and limit bandwidth usage when nothing has changed. Server side JSON response caching is added to limit CPU time.

### Testing steps:
- Open browser devtools and go to http://localhost:3000/extension/redhat/vscode-yaml
- In the devtools you should see `GET | http://localhost:8080/api/redhat/vscode-yaml` returns status 200 and it has a Etag response header.
- Refresh the page. Now `GET | http://localhost:8080/api/redhat/vscode-yaml` has the If-None-Match request header and it returns status 304 (not modified).
- Publish a new version of vscode-yaml. Again refresh the page. `GET | http://localhost:8080/api/redhat/vscode-yaml` returns status 200 again and it has a different Etag value.
- You can repeat the above steps for a specific version too, by using the dropdown to the right.
